Source/NSWindow.m

```diff
@@ -6156,6 +6159,20 @@ - (void) _setWindow: (NSWindow *)window inactive: (BOOL)inactive
 }
 @end
 
+@implementation NSWindow (NSConstraintBasedLayoutCorePrivateMethods)
+
+- (GSAutoLayoutEngine*) _layoutEngine
```

Comment:
You could also add a parameter to  this method, `createIfMissing` or so, and either return nil on missing or create one. That way you could remove the -_setLayoutEngine: and the -_bootstrapAutoLayout method and keep the creation of the GSAutoLayoutEngine in one place.

Reply:
My reasoning for having a separate setter and lazy getter is so I could avoid the primary getter having a side effect,  this has certain tradeoffs as you mentioned.